docker/DOCKER.md

```diff
@@ -0,0 +1,214 @@
+# Docker Images for the balena CLI
+
+Docker images with the balena CLI and docker-in-docker.
+
+## Features Overview
+
+These CLI images are based on the popular [Balena base images](https://www.balena.io/docs/reference/base-images/base-images/)
+so they include many of the features you see there.
+
+- Multiple Architectures:
+    - `rpi`
+    - `armv7hf`
+    - `aarch64` (debian only)
+    - `amd64`
+    - `i386`
+- Multiple Distributions
+    - `debian`
+    - `alpine`
+- [cross-build](https://www.balena.io/docs/reference/base-images/base-images/#building-arm-containers-on-x86-machines) functionality for building ARM containers on x86.
+- Helpful package installer script called `install_packages` inspired by [minideb](https://github.com/bitnami/minideb#why-use-minideb).
+
+Note that there are some additional considerations when running the CLI via Docker so pay close attention to the [Basic usage](#basic-usage) and
+[Advanced Usage](#advanced-usage) sections for examples of different CLI commands.
+
+## Image Names
+
+`balenalib/<arch>-<distro>-balenacli:<cli_ver>`
+
+- `<arch>` is the architecture and is mandatory. If using `Dockerfile.template`, you can replace this with `%%BALENA_ARCH%%`.
+For a list of available device names and architectures, see the [Device types](https://www.balena.io/docs/reference/base-images/devicetypes/).
+- `<distro>` is the Linux distribution and is mandatory. Currently there are 2 distributions, namely `debian` and `alpine`.
+
+## Image Tags
+
+In the tags, all of the fields are optional, and if they are left out, they will default to their `latest` pointer.
+
+- `<cli_ver>` is the version of the balena CLI, for example, `12.40.2`, it can also be substituted for `latest`.
+
+To ensure that the latest available CLI image is used, it is also necessary
+to add the `--pull` flag to `docker run` commands. Otherwise, Docker will
+reuse a locally cached image, without checking for available updates.
+
+## Examples
+
+`balenalib/amd64-debian-balenacli:12.40.2`
+
+- `<arch>`: amd64 - suitable for running on most workstations
+- `<distro>`: debian - widely used base distro
+- `<cli_ver>`: 12.40.2
+
+`balenalib/armv7hf-alpine-balenacli`
+
+- `<arch>`: armv7hf - suitable for running on a Raspberry Pi 3 for example
+- `<distro>`: alpine - smaller footprint than debian
+- `<cli_ver>`: omitted - the latest available CLI version will be used
+
+## Basic Usage
+
+Here's a small example of running a single, detached container
+in the background and using `docker exec` to run balena CLI commands.
+
+```
+$ docker run --pull --detach --privileged --network host --name cli --rm -it balenalib/amd64-debian-balenacli /bin/bash
```

Comment:
I had suggested adding `--pull` here, but it may have to be `--pull always`. Without argument, I get an error if the image does not yet exist locally:
```
$ docker rmi balenalib/amd64-debian-balenacli:1021344afddd361752852d5b4475a6042c767b66

$ docker run --pull --privileged --network host --name cli --rm -it balenalib/amd64-debian-balenacli:1021344afddd361752852d5b4475a6042c767b66
docker: Error response from daemon: No such image: balenalib/amd64-debian-balenacli:1021344afddd361752852d5b4475a6042c767b66.
See 'docker run --help'.
```
```
$ docker run --help
...
      --pull string    Pull image before running ("always"|"missing"|"never") (default "missing")
```
```
$ docker rmi balenalib/amd64-debian-balenacli:1021344afddd361752852d5b4475a6042c767b66

$ docker run --pull always --privileged --network host --name cli --rm -it balenalib/amd64-debian-balenacli:1021344afddd361752852d5b4475a6042c767b66
(success)
```
I am guessing that `--pull` without arguments is taken to mean `--pull never`.

Reply:
Looks like the default without arguments is actually `missing`, which also isn't what we want. I'll change it to `always`.

Reply:
They document `default "missing"`, but the observed behaviour is that:
* No flag: behaves like "missing" (as documented)
* `--pull` provided without argument (which is undocumented usage): behaves like "never"

Reply:
Docker 19.09 is the first version to include the `--pull` option with `docker run` so we may want to recommend a manual pull command rather than a flag that doesn't exist for most versions in use today.

Reply:
Right. Docker v20 is already out, but because of the `preload` AUFS issue we still ask people to downgrade to Docker v18. I believe that removing `--pull always` from the `docker run` command lines will result in people running old versions of the CLI though. There is that note, currently under the Image Tags section:

> To ensure that the latest available CLI image is used, it is also necessary to add the `--pull always` flag to `docker run` commands. Otherwise, Docker will reuse a locally cached image, without checking for available updates.

How about we move that note from the Image Tags section to the Usage section, and reword it a bit, for example as follows:

> We recommend adding the `--pull always` flag to `docker run` commands, so that Docker checks for available updates of the balena CLI image (new image layers will only be downloaded if a new version is available). This is needed even if the `:latest` image tag is used. Note however that `--pull always` is only available with Docker v19 or later. If using an older version of Docker, run a separate `docker pull balenalib/<arch>-<distro>-balenacli` command.

And then we keep the `--pull always` flag in the `docker run` usage examples. If someone is using Docker v18 and `docker run --pull always` fail, they'll easily notice it and hopefully they'll find that note.

Reply:
My version of Docker 19.03 fails as well, the command didn't release until 19.09. What about putting the `docker pull balenalib/amd64-debian-balenacli` before every run command?

Reply:
Shame. Adding the `docker pull` command in every example would be a bit noisy and repetitive. In this case, how about we simply move that note from the Image Tags section to a new section (so it gets more prominence), just before the Usage section, along the lines of:

```
## Keeping the CLI image up to date

Please note that using the `:latest` tag is not enough to keep the image up to date,
because Docker will reuse a locally cached image. To update the image to the latest
version, run:

` ` `bash
$ docker pull balenalib/<arch>-<distro>-balenacli
` ` `

Replacing `<arch>` and `<distro>` with the image architecture and distribution as
described earlier.

If you are using Docker v19.09 or later, you can also add the `--pull always` flag to
`docker run` commands, so that Docker automatically checks for available updates
(new image layers will only be downloaded if a new version is available).
```